File: bindings/python/python/auths/_mechanisms.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Protocol


@dataclass(frozen=True)
class AtomicReservationRecord:
    key: str
    commitment: bytes
    value: bytes


class AtomicReservationStore(Protocol):
    async def reserve(
        self, record: AtomicReservationRecord
    ) -> Literal["acquired", "exact-replay", "conflict"]: ...

    async def aclose(self) -> None: ...

    async def reopen(self) -> AtomicReservationStore: ...

File: bindings/typescript/src/internal/mechanisms.ts
export interface AtomicReservationRecord {
  readonly key: string;
  readonly commitment: Uint8Array;
  readonly value: Uint8Array;
}

export interface AtomicReservationStore {
  reserve(record: AtomicReservationRecord): Promise<"acquired" | "exact-replay" | "conflict">;
  reopen?(): AtomicReservationStore | Promise<AtomicReservationStore>;
  close?(): void | Promise<void>;
}
